src/inumaki/inu_exceptions.py

```diff
@@ -0,0 +1,165 @@
+"""
+Enhanced exception handling for the Inumaki programming language interpreter.
+
+This module provides structured exception classes with descriptive error messages,
+context information, and helpful suggestions for common issues.
+"""
+
+
+class InumakiException(Exception):
+    """Base exception class for all Inumaki interpreter errors."""
+    
+    def __init__(self, message, line=None, column=None, code_snippet=None, suggestion=None):
+        self.message = message
+        self.line = line
+        self.column = column
+        self.code_snippet = code_snippet
+        self.suggestion = suggestion
+        super().__init__(self._format_message())
+    
+    def _format_message(self):
+        """Format the error message with context and suggestions."""
+        parts = [f"{self.__class__.__name__}: {self.message}"]
+        
+        if self.line is not None:
+            parts.append(f"  at line {self.line}" + (f", column {self.column}" if self.column else ""))
+        
+        if self.code_snippet:
+            parts.append(f"  Code: {self.code_snippet}")
+        
+        if self.suggestion:
+            parts.append(f"  Suggestion: {self.suggestion}")
+        
+        return "\n".join(parts)
+
+
+class InumakiSyntaxError(InumakiException):
+    """Raised when there's a syntax error in the source code."""
+    pass
+
+
+class InumakiParseError(InumakiException):
+    """Raised when the parser encounters unexpected tokens or structure."""
+    pass
+
+
+class InumakiRuntimeError(InumakiException):
+    """Raised when there's a runtime error during execution."""
+    pass
+
+
+class InumakiNameError(InumakiRuntimeError):
+    """Raised when trying to access an undefined variable or function."""
+    pass
+
+
+class InumakiTypeError(InumakiRuntimeError):
+    """Raised when there's a type-related error during execution."""
+    pass
+
+
+class InumakiArithmeticError(InumakiRuntimeError):
+    """Raised when there's an arithmetic error like division by zero."""
+    pass
+
+
+class CursedSpeechOverloadError(InumakiRuntimeError):
+    """Raised when the cursed speech threshold is exceeded."""
+    
+    def __init__(self, cursed_count, threshold=100):
```

Comment:
Extract all magic numbers pertaining to the cursed speech threshold and put it as a constant in some sensible location